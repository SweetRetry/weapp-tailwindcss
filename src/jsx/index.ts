import { parse, traverse, generate } from '@/babel'
import type { ASTReplacer } from './replacer'

export function jsxHandler (rawSource: string, replacer: ASTReplacer) {
  // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/53
  replacer?.end()
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  })

  traverse(ast, {
    enter (path) {
      replacer.transform(path)
    },
    noScope: true
  })

  return generate(ast)
}
