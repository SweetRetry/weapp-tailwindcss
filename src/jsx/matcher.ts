import type { Node, ObjectProperty, Identifier, StringLiteral, Expression } from '@babel/types'

export type UserMatchNode = ObjectProperty & { key: Identifier | StringLiteral }

export function isSpecNode (node: Node) {
  return node.type === 'ObjectProperty' && node.key.type === 'Identifier'
}

export function getKey (node: Identifier | StringLiteral | Expression): string {
  if (node.type === 'Identifier') {
    return node.name
  }
  if (node.type === 'StringLiteral') {
    return node.value
  }
  return ''
}

export function reactMatcher (node: UserMatchNode) {
  // rax ->class and className, taro 为 hoverClass remax 为 hoverClassName
  //  node.key.name === 'class' || node.key.name === 'className' || node.key.name === 'hoverClass' || node.key.name === 'hover-class' || node.key.name === 'hoverClassName'
  // react jsx and vue jsx
  return ['className', 'hoverClass', 'hoverClassName', 'class', 'hover-class'].includes(getKey(node.key))
}

export function vue2Matcher (node: UserMatchNode): [boolean, UserMatchNode] {
  // 应该获取指定的节点开始匹配而不是整个 attrs 节点
  // for hover-class
  if (getKey(node.key) === 'attrs' && node.type === 'ObjectProperty') {
    if (node.value.type === 'ObjectExpression' && Array.isArray(node.value.properties)) {
      const idx = node.value.properties.findIndex((x) => {
        return x.type === 'ObjectProperty' && x.key.type === 'StringLiteral' && x.key.value === 'hover-class'
      })
      if (idx > -1) {
        return [true, node.value.properties[idx] as UserMatchNode]
      } else {
        return [false, node]
      }
    }
    return [false, node]
  }
  return [getKey(node.key) === 'class' || getKey(node.key) === 'staticClass', node] // || node.key.name === 'hover-class'
}

export function vue3Matcher (node: ObjectProperty & { key: Identifier | StringLiteral }) {
  if (['class', 'hover-class'].includes(getKey(node.key))) {
    return true
  }
  return false
}

export function isVue3SpecNode (node: Node) {
  return node.type === 'ObjectProperty' && (node.key.type === 'Identifier' || node.key.type === 'StringLiteral')
}
