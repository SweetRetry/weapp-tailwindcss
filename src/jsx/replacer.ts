import type { Node, ObjectProperty, Identifier } from '@babel/types'
import type { NodePath } from '@babel/traverse'
import { replaceWxml } from '@/wxml'
import { classStringReplace } from '@/reg'

export type UserMatchNode = ObjectProperty & { key: Identifier }
export type Replacer = {
  (path: NodePath<Node>): void
  start: (node: Node) => void
  end: () => void
}

function isSpecNode (node: Node) {
  return node.type === 'ObjectProperty' && node.key.type === 'Identifier'
}

function reactMatcher (node: UserMatchNode) {
  return node.key.name === 'className' || node.key.name === 'hoverClass'
}

function vue2Matcher (node: UserMatchNode) {
  return node.key.name === 'class' || node.key.name === 'staticClass' // || node.key.name === 'hover-class'
}

export type JsxFrameworkEnum = 'react' | 'vue' | 'vue2' | 'vue3'

// default react
export function createReplacer (framework: string = 'react'): Replacer {
  let classObjectNode: Node | null
  let startFlag = false
  const isVue3 = framework === 'vue3'
  const isVue2 = framework === 'vue' || framework === 'vue2'
  const isReact = framework === 'react'

  function start (node: Node) {
    startFlag = true
    classObjectNode = node
  }

  function end () {
    startFlag = false
    classObjectNode = null
  }

  if (isVue2) {
    const replacer = (path: NodePath<Node>) => {
      // vue2
      if (isSpecNode(path.node) && vue2Matcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          // TODO
          // 现在这样是有个问题的,变量中用户使用了 'a/s' 就会产生破坏效果
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else if (isVue3) {
    const replacer = (path: NodePath<Node>) => {
      if (
        path.node.type === 'CallExpression' &&
        path.node.arguments.length === 2 &&
        path.node.arguments[0].type === 'StringLiteral' &&
        path.node.arguments[1].type === 'NumericLiteral'
      ) {
        path.node.arguments[0].value = classStringReplace(path.node.arguments[0].value, (x) => {
          return replaceWxml(x, true)
        })
      }

      if (isSpecNode(path.node) && vue2Matcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else if (isReact) {
    const replacer = (path: NodePath<Node>) => {
      // react
      if (isSpecNode(path.node) && reactMatcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else {
    const replacer = (path: NodePath<Node>) => {}
    replacer.start = start
    replacer.end = end
    return replacer
  }
}
