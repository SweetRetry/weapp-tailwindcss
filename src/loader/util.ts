import fs from 'fs'
import path from 'path'

export function mkfileSync (filename: string, content: string) {
  try {
    const dirname = path.dirname(filename)
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, {
        recursive: true
      })
    }
    fs.writeFileSync(filename, content, 'utf-8')
  } catch (error) {
    console.error(error)
  }
}
