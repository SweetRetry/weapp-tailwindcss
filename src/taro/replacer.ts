import type { Node, ObjectProperty, Identifier } from '@babel/types'
import type { NodePath } from '@babel/traverse'
import { replaceWxml } from '../wxml'

export type UserMatchNode = ObjectProperty & { key: Identifier }
export type Replacer = (path: NodePath<Node>) => void

function isSpecNode (node: Node) {
  return node.type === 'ObjectProperty' && node.key.type === 'Identifier'
}

function reactMatcher (node: UserMatchNode) {
  return node.key.name === 'className'
}

function vue2Matcher (node: UserMatchNode) {
  return node.key.name === 'class' || node.key.name === 'staticClass'
}

function vue3Matcher (path: NodePath<Node>) {
  return path.node.type === 'Identifier' && /_hoisted_/.test(path.node.name)
}

export function createReplacer (framework: string = 'react'): Replacer {
  let classObjectNode: Node | null
  let startFlag = false
  const isVue3 = framework === 'vue3'
  const isVue2 = framework === 'vue' || framework === 'vue2'
  const isReact = framework === 'react'
  function end () {
    startFlag = false
    classObjectNode = null
  }

  function start (node: Node) {
    startFlag = true
    classObjectNode = node
  }
  if (isVue2) {
    return function replacer (path: NodePath<Node>) {
      // react and vue2

      // if (vue3Matcher(path)) {
      //   console.log(path.node)
      // }
      if (isSpecNode(path.node) && vue2Matcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          // TODO
          // 现在这样是不恰当的
          // 玩意变量中用户使用了 'a/s' 就会产生破坏效果
          path.node.value = replaceWxml(path.node.value)
        }
      }
    }
  } else if (isVue3) {
    return function replacer (path: NodePath<Node>) {
      // react and vue2

      // if (vue3Matcher(path)) {
      //   console.log(path.node)
      // }
      if (isSpecNode(path.node) && (vue2Matcher(path.node as UserMatchNode) || reactMatcher(path.node as UserMatchNode))) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          // TODO
          // 现在这样是不恰当的
          // 玩意变量中用户使用了 'a/s' 就会产生破坏效果
          path.node.value = replaceWxml(path.node.value)
        }
      }
    }
  } else if (isReact) {
    return function replacer (path: NodePath<Node>) {
      // react and vue2

      // if (vue3Matcher(path)) {
      //   console.log(path.node)
      // }
      if (isSpecNode(path.node) && reactMatcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          // TODO
          // 现在这样是不恰当的
          // 玩意变量中用户使用了 'a/s' 就会产生破坏效果
          path.node.value = replaceWxml(path.node.value)
        }
      }
    }
  } else {
    return function replacer (path: NodePath<Node>) {}
  }
}
