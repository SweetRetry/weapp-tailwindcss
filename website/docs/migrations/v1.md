# 从 v1 迁移到 v2

在 `2.x` 版本中，可以把之前使用的 `webpack` 插件，全部更换为 `UnifiedWebpackPluginV5` 插件，不过 `vite` 插件的导出有一些小变化:

`1.x`:

```js
import vwt from 'weapp-tailwindcss-webpack-plugin/vite';
```

`2.x`:

```js
// UnifiedViteWeappTailwindcssPlugin 就是新的插件
import { UnifiedViteWeappTailwindcssPlugin } from 'weapp-tailwindcss-webpack-plugin/vite';
```

另外新的 `UnifiedWebpackPluginV5` 可以直接从 `weapp-tailwindcss-webpack-plugin` 引入，同时在新的 `UnifiedWebpackPluginV5` 中，之前所有的配置项都被继承了过来，只需要用它直接替换原先插件即可。

另外不要忘记把:

```json
 "scripts": {
+  "postinstall": "weapp-tw patch"
 }
```

添加进你的 `package.json` 里，然后清除原先的打包缓存之后重新打包运行。
