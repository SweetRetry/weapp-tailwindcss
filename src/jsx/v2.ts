import { replaceWxml } from '@/wxml'
import { parse, traverse, generate } from '@/babel'
import type { Node } from '@babel/types'
import { getKey } from './matcher'

const StartMatchKeyMap: Record<'react' | 'vue2' | 'vue3', string[]> = {
  react: ['className', 'hoverClass', 'hoverClassName', 'class', 'hover-class'],
  vue2: ['class', 'staticClass'], // 'hover-class' 在 'attrs' 里
  vue3: ['class', 'hover-class']
}

function isObjectKey (type: string) {
  return ['Identifier', 'StringLiteral'].includes(type)
}

export function newJsxHandler (rawSource: string, framework: keyof typeof StartMatchKeyMap = 'react') {
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  }) as Node
  const matchKeys = StartMatchKeyMap[framework]
  const isVue2 = framework === 'vue2'
  // https://astexplorer.net/
  // TODO
  let startFlag = false

  traverse(ast, {
    ObjectProperty: {
      enter (path) {
        if (isObjectKey(path.node.key.type)) {
          const keyStr = getKey(path.node.key)
          if (matchKeys.includes(keyStr)) {
            startFlag = true
          }
          if (isVue2 && keyStr === 'attrs' && path.node.value.type === 'ObjectExpression') {
            const idx = path.node.value.properties.findIndex((x) => {
              return x.type === 'ObjectProperty' && isObjectKey(x.key.type) && getKey(x.key) === 'hover-class'
            })
            const hoverClassNode = path.node.value.properties[idx]
            if (idx > -1 && hoverClassNode.type === 'ObjectProperty') {
              if (hoverClassNode.value.type === 'StringLiteral') {
                hoverClassNode.value.value = replaceWxml(hoverClassNode.value.value, {
                  keepEOL: true
                })
              }
            }
          }
        }
      },
      exit (path) {
        if (['Identifier', 'StringLiteral'].includes(path.node.key.type) && matchKeys.includes(getKey(path.node.key))) {
          startFlag = false
        }
      }
    },
    StringLiteral: {
      enter (path) {
        if (startFlag) {
          path.node.value = replaceWxml(path.node.value, {
            keepEOL: true
          })
        }
      }
    },
    noScope: true
  })

  return generate(ast)
}
