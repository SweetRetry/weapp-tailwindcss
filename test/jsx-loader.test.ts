import { jsxHandler } from '#test/archived/jsx/v1'
import { newJsxHandler } from '@/jsx/v2'
import { loaderCasePath, createGetCase } from './util'
import { createReplacer } from '#test/archived/jsx/replacer'

const getVue3Case = createGetCase(loaderCasePath + '/taro-vue3-app')
const getVue2Case = createGetCase(loaderCasePath + '/taro-vue2-app')
// const putCase = createPutCase(jsxCasePath)

describe('jsx-loader handler', () => {
  // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/53
  it('vue3 jsx normal case', async () => {
    const code = await getVue3Case('First.tsx.tmp')
    const result = jsxHandler(code, createReplacer('vue3'))
    expect(result).toMatchSnapshot()
  })

  it('new vue3 jsx normal case', async () => {
    const code = await getVue3Case('First.tsx.tmp')
    const result = newJsxHandler(code, {
      framework: 'vue3'
    })
    const result2 = jsxHandler(code, createReplacer('vue3'))
    expect(result).toStrictEqual(result2)
    expect(result).toMatchSnapshot()
  })

  it('new vue2 jsx normal case', async () => {
    const code = await getVue2Case('render.tsx.tmp')
    const result = newJsxHandler(code, {
      framework: 'vue2'
    })
    expect(result).toMatchSnapshot()
  })
})
