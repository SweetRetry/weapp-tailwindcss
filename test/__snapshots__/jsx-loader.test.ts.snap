// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx-loader handler new vue2 jsx normal case 1`] = `
{
  "code": "import Vue from 'vue';
import { View, Button } from '@tarojs/components';
export default Vue.extend({
  components: {
    View: View,
    Button: Button
  },
  render: function render() {
    var h = arguments[0];
    return h(View, {
      "class": 'text-amber-300'
    }, [h(View, {
      "class": "text-center text-_bl_50px_br_ text-_bl__h_123456_br_ mt-_bl_22px_br_",
      "attrs": {
        "hover-class": "text-_bl__h_654321_br_"
      }
    }, ["aaa"]), h(Button, {
      "attrs": {
        "type": 'primary',
        "block": true
      }
    }, ["first"])]);
  }
});",
  "decodedMap": undefined,
  "map": null,
  "rawMappings": undefined,
}
`;

exports[`jsx-loader handler new vue3 jsx normal case 1`] = `
{
  "code": "import { createVNode as _createVNode, createTextVNode as _createTextVNode } from "vue";
import { defineComponent } from 'vue';
import { View, Button } from '@tarojs/components';
export default defineComponent({
  components: {
    View: View,
    Button: Button
  },
  setup: function setup(props, ctx) {
    return function () {
      return _createVNode(View, {
        "class": 'text-amber-300'
      }, {
        default: function _default() {
          return [_createVNode(View, {
            "class": "text-center text-_bl_50px_br_ text-_bl__h_123456_br_ mt-_bl_22px_br_",
            "hover-class": "text-_bl__h_654321_br_"
          }, {
            default: function _default() {
              return [_createTextVNode("components ui")];
            }
          })];
        }
      });
    };
  }
});",
  "decodedMap": undefined,
  "map": null,
  "rawMappings": undefined,
}
`;

exports[`jsx-loader handler vue2 jsx normal case 1`] = `
{
  "code": "import Vue from 'vue';
import { View, Button } from '@tarojs/components';
export default Vue.extend({
  components: {
    View: View,
    Button: Button
  },
  render: function render() {
    var h = arguments[0];
    return h(View, {
      "class": 'text-amber-300'
    }, [h(View, {
      "class": 'text-center text-[50px] text-[#123456] mt-[22px]',
      "attrs": {
        "hover-class": "text-[#654321]"
      }
    }, ["aaa"]), h(Button, {
      "attrs": {
        "type": 'primary',
        "block": true
      }
    }, ["first"])]);
  }
});",
  "decodedMap": undefined,
  "map": null,
  "rawMappings": undefined,
}
`;

exports[`jsx-loader handler vue3 jsx normal case 1`] = `
{
  "code": "import { createVNode as _createVNode, createTextVNode as _createTextVNode } from "vue";
import { defineComponent } from 'vue';
import { View, Button } from '@tarojs/components';
export default defineComponent({
  components: {
    View: View,
    Button: Button
  },
  setup: function setup(props, ctx) {
    return function () {
      return _createVNode(View, {
        "class": 'text-amber-300'
      }, {
        default: function _default() {
          return [_createVNode(View, {
            "class": "text-center text-_bl_50px_br_ text-_bl__h_123456_br_ mt-_bl_22px_br_",
            "hover-class": "text-_bl__h_654321_br_"
          }, {
            default: function _default() {
              return [_createTextVNode("components ui")];
            }
          })];
        }
      });
    };
  }
});",
  "decodedMap": undefined,
  "map": null,
  "rawMappings": undefined,
}
`;

exports[`jsx-loader handler vue3 jsx normal case 2`] = `
{
  "code": "import { createVNode as _createVNode, createTextVNode as _createTextVNode } from "vue";
import { defineComponent } from 'vue';
import { View, Button } from '@tarojs/components';
export default defineComponent({
  components: {
    View: View,
    Button: Button
  },
  setup: function setup(props, ctx) {
    return function () {
      return _createVNode(View, {
        "class": 'text-amber-300'
      }, {
        default: function _default() {
          return [_createVNode(View, {
            "class": "text-center text-_bl_50px_br_ text-_bl__h_123456_br_ mt-_bl_22px_br_",
            "hover-class": "text-_bl__h_654321_br_"
          }, {
            default: function _default() {
              return [_createTextVNode("components ui")];
            }
          })];
        }
      });
    };
  }
});",
  "decodedMap": undefined,
  "map": null,
  "rawMappings": undefined,
}
`;
